### ISSUE TITLE
Segmentation fault on ppc64le

### ISSUE THREAD

**Issue by opoplawski** [h5py/h5py#1164]
To assist reproducing bugs, please include the following:
 * Fedora rawide ppc64le
 * Python version 3.7
 * system python
 * h5py version 2.9.0
 * HDF5 version 1.8.20

```
+ /usr/bin/python3 setup.py test
running test
running build_py
running build_ext
Autodetected HDF5 1.8.20
********************************************************************************
                       Summary of the h5py configuration
    Path to HDF5: None
    HDF5 Version: '1.8.20'
     MPI Enabled: False
Rebuild Required: False
********************************************************************************
Executing cythonize()
BUILDSTDERR: .FFFF......s.................F.........../var/tmp/rpm-tmp.hAtZZ4: line 31: 26868 Segmentation fault      (core dumped) /usr/bin/python3 setup.py test
```
This is new with 2.9.0 - I don't see it with 2.8.0.  I'll work on getting a backtrace.

**Comment by opoplawski** [comment]
Well, when I go to run the tests directly (outside of the mock build) I get different output:
```
$ /usr/bin/python3 setup.py test -v
running test
running build_py
running build_ext
Autodetected HDF5 1.8.20
********************************************************************************
                       Summary of the h5py configuration

    Path to HDF5: None
    HDF5 Version: '1.8.20'
     MPI Enabled: False
Rebuild Required: False

********************************************************************************
Executing cythonize()
.FFFF......s.................F...........F.../builddir/build/BUILD/h5py-2.9.0/build/lib.linux-ppc64le-3.7/h5py/tests/old/test_dataset.py:888: FutureWarning: elementwise == comparison failed and returning scalar instead; this will raise an error or perform elementwise comparison in the future.
  self.assertTrue(np.all(outdata == testdata))
F/builddir/build/BUILD/h5py-2.9.0/build/lib.linux-ppc64le-3.7/h5py/tests/old/test_dataset.py:871: FutureWarning: elementwise == comparison failed and returning scalar instead; this will raise an error or perform elementwise comparison in the future.
  self.assertTrue(np.all(outdata == testdata))
F........x....................F.......E...............F...........................x........FF..........................s/builddir/build/BUILD/h5py-2.9.0/build/lib.linux-ppc64le-3.7/h5py/_hl/files.py:258: H5pyDeprecationWarning: File.fid has been deprecated. Use File.id instead.
  "Use File.id instead.", H5pyDeprecationWarning)
...s......ss.......................................................................................................ssssss................................E..........F..F............/usr/lib64/python3.7/unittest/case.py:846: FutureWarning: elementwise == comparison failed and returning scalar instead; this will raise an error or perform elementwise comparison in the future.
  if not first == second:
F.........x....x.........................x......x........................................ssssssssss..............................E..FF....sssssssssssssssssssssssss
======================================================================
ERROR: test_float (h5py.tests.old.test_dataset.TestCreateScaleOffset)
Scaleoffset filter works for floating point data
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builddir/build/BUILD/h5py-2.9.0/build/lib.linux-ppc64le-3.7/h5py/tests/old/test_dataset.py", line 450, in test_float
    dset = self.f.create_dataset('foo', shape, dtype=float, scaleoffset=scalefac)
  File "/builddir/build/BUILD/h5py-2.9.0/build/lib.linux-ppc64le-3.7/h5py/_hl/group.py", line 136, in create_dataset
    dsid = dataset.make_new_dset(self, shape, dtype, data, **kwds)
  File "/builddir/build/BUILD/h5py-2.9.0/build/lib.linux-ppc64le-3.7/h5py/_hl/dataset.py", line 167, in make_new_dset
    dset_id = h5d.create(parent.id, None, tid, sid, dcpl=dcpl)
  File "h5py/_objects.pyx", line 54, in h5py._objects.with_phil.wrapper
    with _phil:
  File "h5py/_objects.pyx", line 55, in h5py._objects.with_phil.wrapper
    return func(*args, **kwds)
  File "h5py/h5d.pyx", line 79, in h5py.h5d.create
    dsid = H5Dcreate_anon(loc.id, tid.id, space.id,
ValueError: Unable to create dataset (cannot find matched memory dataype)

======================================================================
ERROR: test_out_of_order_offsets (h5py.tests.old.test_h5t.TestCompound)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builddir/build/BUILD/h5py-2.9.0/build/lib.linux-ppc64le-3.7/h5py/tests/old/test_h5t.py", line 59, in test_out_of_order_offsets
    h5t.py_create(dt)
  File "h5py/_objects.pyx", line 54, in h5py._objects.with_phil.wrapper
    with _phil:
  File "h5py/_objects.pyx", line 55, in h5py._objects.with_phil.wrapper
    return func(*args, **kwds)
  File "h5py/h5t.pyx", line 1172, in h5py.h5t.TypeCompoundID.insert
    H5Tinsert(self.id, name, offset, field.id)
ValueError: Unable to insert member (member overlaps with another member)

======================================================================
ERROR: test_aligned_data (h5py.tests.hl.test_datatype.TestOffsets)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builddir/build/BUILD/h5py-2.9.0/build/lib.linux-ppc64le-3.7/h5py/tests/hl/test_datatype.py", line 195, in test_aligned_data
    f['data'] = data
  File "/builddir/build/BUILD/h5py-2.9.0/build/lib.linux-ppc64le-3.7/h5py/_hl/group.py", line 385, in __setitem__
    ds = self.create_dataset(None, data=obj, dtype=base.guess_dtype(obj))
  File "/builddir/build/BUILD/h5py-2.9.0/build/lib.linux-ppc64le-3.7/h5py/_hl/group.py", line 136, in create_dataset
    dsid = dataset.make_new_dset(self, shape, dtype, data, **kwds)
  File "/builddir/build/BUILD/h5py-2.9.0/build/lib.linux-ppc64le-3.7/h5py/_hl/dataset.py", line 118, in make_new_dset
    tid = h5t.py_create(dtype, logical=1)
  File "h5py/h5t.pyx", line 1630, in h5py.h5t.py_create
    cpdef TypeID py_create(object dtype_in, bint logical=0, bint aligned=0):
  File "h5py/h5t.pyx", line 1652, in h5py.h5t.py_create
    with phil:
  File "h5py/h5t.pyx", line 1674, in h5py.h5t.py_create
    return _c_compound(dt, logical, aligned)
  File "h5py/h5t.pyx", line 1590, in h5py.h5t._c_compound
    raise TypeError("Enforced alignment not compatible with HDF5 type")
TypeError: Enforced alignment not compatible with HDF5 type

======================================================================
FAIL: test_create (h5py.tests.old.test_attrs.TestAccess)
Attribute creation by direct assignment
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builddir/build/BUILD/h5py-2.9.0/build/lib.linux-ppc64le-3.7/h5py/tests/old/test_attrs.py", line 53, in test_create
    self.assertEqual(self.f.attrs['a'], 4.0)
AssertionError: 4.0 != 4.0

======================================================================
FAIL: test_overwrite (h5py.tests.old.test_attrs.TestAccess)
Attributes are silently overwritten
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builddir/build/BUILD/h5py-2.9.0/build/lib.linux-ppc64le-3.7/h5py/tests/old/test_attrs.py", line 59, in test_overwrite
    self.assertEqual(self.f.attrs['a'], 5.0)
AssertionError: 5.0 != 5.0

======================================================================
FAIL: test_rank (h5py.tests.old.test_attrs.TestAccess)
Attribute rank is preserved
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builddir/build/BUILD/h5py-2.9.0/build/lib.linux-ppc64le-3.7/h5py/tests/old/test_attrs.py", line 65, in test_rank
    self.assertArrayEqual(self.f.attrs['a'], np.array((4.0,5.0)))
  File "/builddir/build/BUILD/h5py-2.9.0/build/lib.linux-ppc64le-3.7/h5py/tests/common.py", line 124, in assertArrayEqual
    "Dtype mismatch (%s vs %s)%s" % (dset.dtype, arr.dtype, message)
AssertionError: False is not true : Dtype mismatch (float128 vs float64)

======================================================================
FAIL: test_single (h5py.tests.old.test_attrs.TestAccess)
Attributes of shape (1,) don't become scalars
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builddir/build/BUILD/h5py-2.9.0/build/lib.linux-ppc64le-3.7/h5py/tests/old/test_attrs.py", line 72, in test_single
    self.assertEqual(out[()], 1)
AssertionError: array([1.], dtype=float128) != 1

======================================================================
FAIL: test_float (h5py.tests.old.test_attrs_data.TestTypes)
Storage of floating point types
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builddir/build/BUILD/h5py-2.9.0/build/lib.linux-ppc64le-3.7/h5py/tests/old/test_attrs_data.py", line 112, in test_float
    self.assertEqual(out.dtype, dt)
AssertionError: dtype('<f16') != dtype('float64')

======================================================================
FAIL: test_astype (h5py.tests.old.test_dataset.TestAstype)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builddir/build/BUILD/h5py-2.9.0/build/lib.linux-ppc64le-3.7/h5py/tests/old/test_dataset.py", line 1021, in test_astype
    self.assertTrue(np.all(dset[...] == np.arange(100)))
AssertionError: False is not true

======================================================================
FAIL: test_assign (h5py.tests.old.test_dataset.TestCompound)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builddir/build/BUILD/h5py-2.9.0/build/lib.linux-ppc64le-3.7/h5py/tests/old/test_dataset.py", line 888, in test_assign
    self.assertTrue(np.all(outdata == testdata))
AssertionError: False is not true

======================================================================
FAIL: test_rt (h5py.tests.old.test_dataset.TestCompound)
Compound types are read back in correct order (issue 236)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builddir/build/BUILD/h5py-2.9.0/build/lib.linux-ppc64le-3.7/h5py/tests/old/test_dataset.py", line 871, in test_rt
    self.assertTrue(np.all(outdata == testdata))
AssertionError: False is not true

======================================================================
FAIL: test_named (h5py.tests.old.test_dataset.TestCreateNamedType)
Named type object works and links the dataset to type
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builddir/build/BUILD/h5py-2.9.0/build/lib.linux-ppc64le-3.7/h5py/tests/old/test_dataset.py", line 282, in test_named
    self.assertEqual(dset.dtype, np.dtype('f8'))
AssertionError: dtype('<f16') != dtype('float64')

======================================================================
FAIL: test_external (h5py.tests.old.test_dataset.TestExternal)
Create and access an external dataset
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builddir/build/BUILD/h5py-2.9.0/build/lib.linux-ppc64le-3.7/h5py/tests/old/test_dataset.py", line 553, in test_external
    assert statinfo.st_size == testdata.nbytes
AssertionError

======================================================================
FAIL: test_numpy_float64_2 (h5py.tests.old.test_dataset.TestVlen)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builddir/build/BUILD/h5py-2.9.0/build/lib.linux-ppc64le-3.7/h5py/tests/old/test_dataset.py", line 1158, in test_numpy_float64_2
    self._help_float_testing(np_dt)
  File "/builddir/build/BUILD/h5py-2.9.0/build/lib.linux-ppc64le-3.7/h5py/tests/old/test_dataset.py", line 1127, in _help_float_testing
    self.assertArrayEqual(array_0, ds[0])
  File "/builddir/build/BUILD/h5py-2.9.0/build/lib.linux-ppc64le-3.7/h5py/tests/common.py", line 124, in assertArrayEqual
    "Dtype mismatch (%s vs %s)%s" % (dset.dtype, arr.dtype, message)
AssertionError: False is not true : Dtype mismatch (float64 vs float128)

======================================================================
FAIL: test_numpy_float64_from_dtype (h5py.tests.old.test_dataset.TestVlen)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builddir/build/BUILD/h5py-2.9.0/build/lib.linux-ppc64le-3.7/h5py/tests/old/test_dataset.py", line 1154, in test_numpy_float64_from_dtype
    self._help_float_testing(np_dt)
  File "/builddir/build/BUILD/h5py-2.9.0/build/lib.linux-ppc64le-3.7/h5py/tests/old/test_dataset.py", line 1127, in _help_float_testing
    self.assertArrayEqual(array_0, ds[0])
  File "/builddir/build/BUILD/h5py-2.9.0/build/lib.linux-ppc64le-3.7/h5py/tests/common.py", line 124, in assertArrayEqual
    "Dtype mismatch (%s vs %s)%s" % (dset.dtype, arr.dtype, message)
AssertionError: False is not true : Dtype mismatch (float64 vs float128)

======================================================================
FAIL: test_read (h5py.tests.old.test_slicing.TestArraySlicing)
Read arrays tack array dimensions onto end of shape tuple
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builddir/build/BUILD/h5py-2.9.0/build/lib.linux-ppc64le-3.7/h5py/tests/old/test_slicing.py", line 165, in test_read
    self.assertEqual(dset.dtype, dt)
AssertionError: dtype(('<f16', (3,))) != dtype(('<f8', (3,)))

======================================================================
FAIL: test_write_element (h5py.tests.old.test_slicing.TestArraySlicing)
Write a single element to the array
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builddir/build/BUILD/h5py-2.9.0/build/lib.linux-ppc64le-3.7/h5py/tests/old/test_slicing.py", line 204, in test_write_element
    self.assertTrue(np.all(out == data))
AssertionError: False is not true

======================================================================
FAIL: test_compound (h5py.tests.old.test_slicing.TestSingleElement)
Compound scalar is numpy.void, not tuple (issue 135)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builddir/build/BUILD/h5py-2.9.0/build/lib.linux-ppc64le-3.7/h5py/tests/old/test_slicing.py", line 77, in test_compound
    self.assertEqual(dset[0], v[0])
AssertionError: (1, 1.) != (1, 1.)

======================================================================
FAIL: test_float_round_tripping (h5py.tests.hl.test_datatype.TestOffsets)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builddir/build/BUILD/h5py-2.9.0/build/lib.linux-ppc64le-3.7/h5py/tests/hl/test_datatype.py", line 240, in test_float_round_tripping
    self.assertEqual(ldata.dtype, d)
AssertionError: dtype('<f16') != <class 'numpy.float64'>

======================================================================
FAIL: test_out_of_order_offsets (h5py.tests.hl.test_datatype.TestOffsets)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builddir/build/BUILD/h5py-2.9.0/build/lib.linux-ppc64le-3.7/h5py/tests/hl/test_datatype.py", line 217, in test_out_of_order_offsets
    self.assertArrayEqual(fd['data'], data)
  File "/builddir/build/BUILD/h5py-2.9.0/build/lib.linux-ppc64le-3.7/h5py/tests/common.py", line 124, in assertArrayEqual
    "Dtype mismatch (%s vs %s)%s" % (dset.dtype, arr.dtype, message)
AssertionError: False is not true : Dtype mismatch ({'names':['f1','f2','f3'], 'formats':['<f4','<i4','<f16'], 'offsets':[0,24,8], 'itemsize':28} vs {'names':['f1','f2','f3'], 'formats':['<f4','<i4','<f8'], 'offsets':[0,16,8], 'itemsize':20})

----------------------------------------------------------------------
Ran 509 tests in 0.680s

FAILED (failures=17, errors=3, skipped=46, expected failures=6)
corrupted size vs. prev_size
Aborted (core dumped)
```

**Comment by tacaswell** [comment]
My guess is that this is due to https://github.com/h5py/h5py/pull/1114 or https://github.com/h5py/h5py/pull/1065 sorry :disappointed: .

I do not have a machine with the correct architecture to debug this on...

**Comment by opoplawski** [comment]
I know it's not convenient debugging via a third-part - but if there are any tests that you would like me to run on ppc64le let me know.

**Comment by tacaswell** [comment]
Can you try reverting those two PRs locally and see if it works again?  Failing that, put lots of prints in the type mapping code to see if we can sort out where that is going wrong....

**Comment by cdluminate** [comment]
The same problem for Debian: https://buildd.debian.org/status/fetch.php?pkg=h5py&arch=ppc64el&ver=2.9.0-1&stamp=1549098779&raw=0

**Comment by takluyver** [comment]
Closed this as I believe it was fixed by PR #1243. There are apparently test failures still remaining - there should be a separate issue describing those soon. But the segfault should be fixed, at least.

As far as I know, we're reliant on people with access to ppc64le systems to find and resolve issues like this, so do keep testing.
